src/unit_selection.rs

```diff
@@ -0,0 +1,143 @@
+use bevy::prelude::*;
+use bevy_mod_picking::*;
+
+#[derive(Debug)]
+pub struct SelectionPlugin;
+
+impl Plugin for SelectionPlugin
+{
+    fn build(&self, app: &mut AppBuilder)
+    {
+        app
+            .add_event::<SelectionEvent>()
+            .init_resource::<SelectionState>()
+            .init_resource::<SelectionColor>()
+            .init_resource::<SelectionEventReader>()
+            .add_system(selection_update.system())
+            .add_system(highlight_selected.system());
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct Selectable;
+
+#[derive(Debug, Clone)]
+pub struct SelectionEvent {
+    selected: Entity
+}
+
+#[derive(Default)]
+pub struct SelectionEventReader {
+    reader: EventReader<SelectionEvent>
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct SelectionState
+{
+    pub currently_selected: Option<Entity>,
+    pub last_selected: Option<Entity>,
+}
+
+impl SelectionState
+{
+    /// Swaps the currently selected to the last selected, and clears the currently selected.
+    pub fn swap_and_clear(&mut self)
+    {
+        self.current_to_last();
+        self.currently_selected = None;
+    }
+
+    /// Sets the currently selected to the last selected.
+    pub fn current_to_last(&mut self)
+    {
+        self.last_selected = self.currently_selected;
+    }
+}
+
+pub struct SelectionColor(pub Color);
+
+impl SelectionColor
+{
+    const SELECTED_COLOR: Color = Color::rgb(0.3, 0.8, 0.5);
+}
+
+impl Default for SelectionColor
+{
+    fn default() -> Self {
+        SelectionColor(Self::SELECTED_COLOR)
+    }
+}
+
+
+fn selection_update(
+    pick_state: Res<PickState>,
+    mut selection_state: ResMut<SelectionState>,
+    mut selection_events: ResMut<Events<SelectionEvent>>,
+    mouse_button_inputs: Res<Input<MouseButton>>,
+    query: Query<&Selectable>
+)
+{
+    if mouse_button_inputs.just_pressed(MouseButton::Left)
+    {
+        if let Some(pick_depth) = pick_state.top()
+        {
+            selection_state.swap_and_clear();
+            
+            if let Ok(_) = query.get::<Selectable>(pick_depth.entity())
```

Comment:
```suggestion
            if query.get::<Selectable>(pick_depth.entity()).is_ok()
```
Unnecessary pattern matching is reducing the readability.

Reply:
This is really a style thing honestly, I think pattern matching is a pretty readable way of expressing it :P

But I'll probably change it anyways.